File: python/problem_24.py
class LockNode:
    def __init__(self, val, parent=None):
        self.val = val
        self.locked = False
        self.locked_desc_count = 0
        self.parent = parent
        self.left = None
        self.right = None

    def is_locked(self):
        return self.locked

    def _can_lock_or_unlock(self):
        if self.locked_desc_count > 0:
            return False
        p = self.parent
        while p:
            if p.locked:
                return False
            p = p.parent
        return True

    def lock(self):
        if self.locked or not self._can_lock_or_unlock():
            return False
        self.locked = True
        p = self.parent
        while p:
            p.locked_desc_count += 1
            p = p.parent
        return True

    def unlock(self):
        if not self.locked or not self._can_lock_or_unlock():
            return False
        self.locked = False
        p = self.parent
        while p:
            p.locked_desc_count -= 1
            p = p.parent
        return True


def main():
    print("Problem 24:")
    root = LockNode(1)
    root.left = LockNode(2, root)
    root.right = LockNode(3, root)
    root.left.left = LockNode(4, root.left)

    assert root.left.lock() == True
    assert root.left.is_locked() == True
    assert root.lock() == False              # has locked descendant
    assert root.left.left.lock() == False    # ancestor locked
    assert root.left.unlock() == True
    assert root.left.is_locked() == False
    assert root.lock() == True
    print("All Tests Passed")

File: rust/problem_24.rs
// Locking binary tree. Nodes are stored in a flat Vec; indices act as pointers.
// Each node tracks parent index and a count of locked descendants.
// NULL sentinel: usize::MAX.

pub struct LockTree {
    vals:              Vec<i64>,
    locked:            Vec<bool>,
    locked_desc_count: Vec<usize>,
    parent:            Vec<usize>,
    left:              Vec<usize>,
    right:             Vec<usize>,
}

impl LockTree {
    pub fn new() -> Self {
        LockTree {
            vals:              Vec::new(),
            locked:            Vec::new(),
            locked_desc_count: Vec::new(),
            parent:            Vec::new(),
            left:              Vec::new(),
            right:             Vec::new(),
        }
    }

    pub fn add_node(&mut self, val: i64, parent: usize) -> usize {
        let idx = self.vals.len();
        self.vals.push(val);
        self.locked.push(false);
        self.locked_desc_count.push(0);
        self.parent.push(parent);
        self.left.push(usize::MAX);
        self.right.push(usize::MAX);
        idx
    }

    pub fn is_locked(&self, idx: usize) -> bool { self.locked[idx] }

    fn can_lock_or_unlock(&self, idx: usize) -> bool {
        if self.locked_desc_count[idx] > 0 { return false; }
        let mut p = self.parent[idx];
        while p != usize::MAX {
            if self.locked[p] { return false; }
            p = self.parent[p];
        }
        true
    }

    pub fn lock(&mut self, idx: usize) -> bool {
        if self.locked[idx] || !self.can_lock_or_unlock(idx) { return false; }
        self.locked[idx] = true;
        let mut p = self.parent[idx];
        while p != usize::MAX { self.locked_desc_count[p] += 1; p = self.parent[p]; }
        true
    }

    pub fn unlock(&mut self, idx: usize) -> bool {
        if !self.locked[idx] || !self.can_lock_or_unlock(idx) { return false; }
        self.locked[idx] = false;
        let mut p = self.parent[idx];
        while p != usize::MAX { self.locked_desc_count[p] -= 1; p = self.parent[p]; }
        true
    }
}

pub fn problem24() {
    println!("Problem 24:");
    let mut t = LockTree::new();
    let root  = t.add_node(1, usize::MAX);
    let left  = t.add_node(2, root);
    let right = t.add_node(3, root);
    let ll    = t.add_node(4, left);
    t.left[root] = left; t.right[root] = right; t.left[left] = ll;

    assert!(t.lock(left),         "should lock left");
    assert!(t.is_locked(left),    "left should be locked");
    assert!(!t.lock(root),        "root has locked descendant");
    assert!(!t.lock(ll),          "ll has locked ancestor");
    assert!(t.unlock(left),       "should unlock left");
    assert!(!t.is_locked(left),   "left should be unlocked");
    assert!(t.lock(root),         "root should now lock");
    println!("All Tests Passed");
}
